jbmc/unit/java_bytecode/java_trace_validation/java_trace_validation.cpp

```diff
@@ -0,0 +1,128 @@
+    address_of_exprt(constant_exprt("0", java_int_type()));
+  const struct_exprt struct_plain =
+    struct_exprt(std::vector<exprt>(), java_int_type());
+  const array_exprt array_plain =
+    array_exprt(std::vector<exprt>(), array_typet(java_int_type(), exprt()));
+  const array_list_exprt array_list_plain = array_list_exprt(
+    std::vector<exprt>(), array_typet(java_int_type(), exprt()));
+
+  SECTION("check_symbol_structure")
+  {
+    INFO("valid symbol expression")
+    REQUIRE(check_symbol_structure(valid_symbol_expr));
+    INFO("invalid symbol expression, missing identifier")
+    REQUIRE_FALSE(check_symbol_structure(invalid_symbol_expr));
+    INFO("invalid symbol expression, not a symbol")
+    REQUIRE_FALSE(check_symbol_structure(plain_expr));
+  }
+
+  SECTION("get_inner_symbol_expr")
+  {
+    const exprt inner_symbol = exprt(exprt(valid_symbol_expr));
+    const exprt inner_nonsymbol = exprt(exprt(exprt()));
+    INFO("expression does not have an inner symbol")
+    REQUIRE(get_inner_symbol_expr(inner_symbol).has_value());
+    INFO("expression does not have an inner symbol")
+    REQUIRE_FALSE(get_inner_symbol_expr(inner_nonsymbol).has_value());
+  }
+
+  SECTION("check_member_structure")
+  {
+    INFO("valid member structure")
+    REQUIRE(check_member_structure(valid_member));
+    INFO("not a member")
+    REQUIRE_FALSE(check_member_structure(plain_expr));
+    INFO("invalid member structure, no symbol operand")
+    REQUIRE_FALSE(check_member_structure(invalid_member));
+  }
+
+  SECTION("valid_lhs_expr_high_level")
+  {
+    INFO("member_exprts are valid lhs expressions")
+    REQUIRE(valid_lhs_expr_high_level(valid_member));
+    INFO("symbol_exprts are valid lhs expressions")
+    REQUIRE(valid_lhs_expr_high_level(valid_symbol_expr));
+    INFO("index_exprts are valid lhs expressions")
+    REQUIRE(valid_lhs_expr_high_level(index_plain));
+    INFO("byte_extract_exprts little endian are valid lhs expressions")
+    REQUIRE(valid_lhs_expr_high_level(byte_little_endian));
+    INFO("address_of_exprts are not valid lhs expressions, for example")
+    REQUIRE_FALSE(valid_lhs_expr_high_level(valid_address));
+  }
+
+  SECTION("valid_rhs_expr_high_level")
+  {
+    INFO("symbol_exprts are valid rhs expressions")
+    REQUIRE(valid_rhs_expr_high_level(valid_symbol_expr));
+    INFO("address_of_exprts are valid lhs expressions")
+    REQUIRE(valid_rhs_expr_high_level(valid_address));
+    INFO("struct_exprts are valid lhs expressions")
+    REQUIRE(valid_rhs_expr_high_level(struct_plain));
+    INFO("array_exprts are valid lhs expressions")
+    REQUIRE(valid_rhs_expr_high_level(array_plain));
+    INFO("array_list_exprts are valid lhs expressions")
+    REQUIRE(valid_rhs_expr_high_level(array_list_plain));
+    INFO("constant_exprts are valid lhs expressions")
+    REQUIRE(valid_rhs_expr_high_level(valid_constant));
+    INFO("member_exprts are not valid lhs expressions, for example")
```

Comment:
:question: should this read `member_exprts` are not valid *rhs* expression? since we're in the rhs? :confused: